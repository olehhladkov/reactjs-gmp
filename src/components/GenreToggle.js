import { useSelector, useDispatch } from 'react-redux';
import { getMoviesThunk, setFilterOption } from '../app/appSlice';
import '../styles/GenreToggle.scss';

function GenreToggle() {
  const filterOptionList = useSelector((state) => state.app.filterOptionList);
  const filterOptionSelected = useSelector(
    (state) => state.app.filterOptionSelected
  );
  const sortOptionSelected = useSelector(
    (state) => state.app.sortOptionSelected
  );

  const dispatch = useDispatch();

  const applyFilter = (filterOption) => {
    dispatch(setFilterOption(filterOption));
    dispatch(
      getMoviesThunk(
        `?filter=${filterOption}&sortBy=${sortOptionSelected.value}`
      )
    );
  };

  return (
    <ul className="genre-toggle">
      {filterOptionList.map((genre) => {
        return (
          <li
            key={genre}
            className={`genre-toggle__item ${
              genre === filterOptionSelected ? 'active' : ''
            }`}
          >
            <button
              className="genre-toggle__btn"
              type="button"
              disabled={genre === filterOptionSelected}
              onClick={() => applyFilter(genre)}
            >
              {genre || 'All'}
            </button>
          </li>
        );
      })}
    </ul>
  );
}

export default GenreToggle;
