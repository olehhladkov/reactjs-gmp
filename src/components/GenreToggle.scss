@import 'colors.scss';

.wrapper {
  padding: 1rem 0;
  background-color: $dark;
}

.genre-toggle {
  display: flex;
  gap: 0 2rem;
  max-width: 62.5rem; //1000px
  margin: 0 auto;
  padding: 0;
  color: $light;
  list-style: none;
  border-bottom: 1px solid $dark-400;

  &__item {
    border-top: 2px solid transparent;
    border-bottom: 2px solid transparent;

    &.active {
      border-bottom-color: $accent;
    }
  }

  &__btn {
    padding: 1.25rem 0;
    font-size: inherit;
    text-transform: uppercase;
    color: inherit;
    border: 0;
    background-color: transparent;
    cursor: pointer;
  }
}
