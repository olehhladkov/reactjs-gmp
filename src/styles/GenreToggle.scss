.genre-toggle {
  display: flex;
  gap: 0 2rem;
  margin: 0 0 -1px;
  padding: 0;
  list-style: none;

  &__item {
    border-top: 2px solid transparent;
    border-bottom: 2px solid transparent;

    &.active {
      border-bottom-color: var(--accent);
    }
  }

  &__btn {
    padding: 1.25rem 0;
    font-size: inherit;
    font-weight: 500;
    text-transform: uppercase;
    color: var(--light);
    border: 0;
    background-color: transparent;

    &:not([disabled]) {
      cursor: pointer;

      &:hover {
        color: var(--accent-hover);
      }
    }
  }
}
