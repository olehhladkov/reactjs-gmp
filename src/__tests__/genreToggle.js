import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import GenreToggle from '../components/GenreToggle';
import reducer from '../app/appSlice';

describe('GenreToggle', () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: { app: reducer },
    });
  });

  it('renders correctly', async () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <GenreToggle />
        </Provider>
      </MemoryRouter>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('handles click event', async () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <GenreToggle />
        </Provider>
      </MemoryRouter>
    );
    const button = container.querySelector('li:nth-child(2) button');
    const { textContent } = button;

    fireEvent.click(button);

    expect(store.getState().app.searchFields.genre).toEqual(textContent);
  });
});
