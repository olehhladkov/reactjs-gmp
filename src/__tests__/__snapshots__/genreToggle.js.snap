// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreToggle renders correctly 1`] = `
<ul
  class="genre-toggle"
>
  <li
    class="genre-toggle__item active"
  >
    <button
      class="genre-toggle__btn"
      disabled=""
      type="button"
    >
      All
    </button>
  </li>
  <li
    class="genre-toggle__item "
  >
    <button
      class="genre-toggle__btn"
      type="button"
    >
      comedy
    </button>
  </li>
  <li
    class="genre-toggle__item "
  >
    <button
      class="genre-toggle__btn"
      type="button"
    >
      crime
    </button>
  </li>
  <li
    class="genre-toggle__item "
  >
    <button
      class="genre-toggle__btn"
      type="button"
    >
      documentary
    </button>
  </li>
  <li
    class="genre-toggle__item "
  >
    <button
      class="genre-toggle__btn"
      type="button"
    >
      drama
    </button>
  </li>
  <li
    class="genre-toggle__item "
  >
    <button
      class="genre-toggle__btn"
      type="button"
    >
      horror
    </button>
  </li>
</ul>
`;
